wsmux/stream.go

```diff
@@ -0,0 +1,318 @@
+	DefaultCapacity = 1024
+)
+
+/*Stream States:
+created = stream has been created. Buffer is empty. Has not been accepted.
+accepted = stream has been accepted. read write operations permitted.
+closed = stream has been closed.
+remoteClosed = remote side has been closed.
+dead = closed & remoteClosed. Buffer may still have data.
+*/
+type streamState int
+
+const (
+	created streamState = iota
+	accepted
+	closed
+	remoteClosed
+	dead
+)
+
+type stream struct {
+	id uint32     // id of the stream. Used for logging.
+	m  sync.Mutex // mutex for state transitions
+	c  *sync.Cond // used for broadcasting when closed, data read, or data pushed to buffer
+	b  *buffer    // read buffer
+
+	unblocked uint32 // number of bytes that can be sent to remote
+
+	endErr   error         // error causes stream to close
+	state    streamState   // current state of the stream
+	accepted chan struct{} // closed when stream is accepted. Used in session.Open()
+
+	session *Session // assosciated session. used for sending frames and logging
+
+	readTimer             *time.Timer // timer for read operations
+	writeTimer            *time.Timer // timer for write operations
+	readDeadlineExceeded  bool        // true when readTimer fires
+	writeDeadlineExceeded bool        // true when writeTimer fires
+}
+
+func newStream(id uint32, session *Session) *stream {
+	str := &stream{
+		id:        id,
+		b:         newBuffer(DefaultCapacity),
+		unblocked: 0,
+		state:     created,
+		accepted:  make(chan struct{}),
+
+		endErr: nil,
+
+		readTimer:             nil,
+		writeTimer:            nil,
+		readDeadlineExceeded:  false,
+		writeDeadlineExceeded: false,
+
+		session: session,
+	}
+
+	str.c = sync.NewCond(&str.m)
+
+	return str
+}
+
+// HandleFrame processes frames received by the stream
+func (s *stream) HandleFrame(fr frame) {
+	switch fr.msg {
+	case msgACK:
+		// if not accepted then close accepted channel, otherwise unblock
+		read := binary.LittleEndian.Uint32(fr.payload)
+		select {
+		case <-s.accepted:
+			s.UnblockAndBroadcast(read)
+		default:
+			s.AcceptStream(read)
+		}
+	case msgDAT:
+		s.session.logger.Printf("stream %d received DAT frame: %v", s.id, fr)
+		s.PushAndBroadcast(fr.payload)
+	case msgFIN:
+		s.session.logger.Printf("remote stream %d closed connection", s.id)
+		s.setRemoteClosed()
+	}
+}
+
+// onExpired is an internal helper method which sets val = true and broadcasts
+func (s *stream) onExpired(val *bool) func() {
+	return func() {
+		s.m.Lock()
+		defer s.m.Unlock()
+		defer s.c.Broadcast()
+		*val = true
+	}
+}
+
+// SetReadDeadline sets the read timer
+func (s *stream) SetReadDeadline(t time.Time) error {
+
+	s.m.Lock()
+	defer s.m.Unlock()
+	// stop timer if not nil
+	if s.readTimer != nil {
+		_ = s.readTimer.Stop()
+		s.readTimer = nil
+	}
+	// clear deadline exceeded
+	s.readDeadlineExceeded = false
+	if !t.IsZero() {
+		delay := t.Sub(time.Now())
+		s.readTimer = time.AfterFunc(delay, s.onExpired(&s.readDeadlineExceeded))
```

Comment:
I suspect reusing the timer is more efficient... You certainly reduce the number of heap allocations...
Given that this moves for every single chunk it might be worth considering...